apps/app_utf8.c

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <openssl/buffer.h>
+#include <openssl/asn1.h>        /* For UTF8_putc */
+#include "apps.h"
+
+char *to_utf8(const char *str)
+{
+    char *str_utf8 = NULL;
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199409L \
+    || defined(_WIN32)
+    wchar_t *wstr = NULL;
+    size_t wstr_size;
+    size_t utf8_count, i, j;
+
+    /*
+     * Because MultiByteToWideChar returns 0 for error,
+     * avoid trying to convert empty strings
+     */
+    if (str[0] == '\0')
+        return OPENSSL_strdup("");
+
+# ifdef _WIN32
+    if (GetEnvironmentVariableW(L"OPENSSL_WIN32_UTF8", NULL, 0) != 0) {
+        return OPENSSL_strdup(str);
+
+    if ((wstr_size = MultiByteToWideChar(CP_ACP,
+                                         MB_ERR_INVALID_CHARS | MB_PRECOMPOSED,
+                                         str, -1, NULL, 0)) == 0
+        || (wstr = OPENSSL_malloc(wstr_size * sizeof(*wstr))) == NULL)
+        return NULL;
+    MultiByteToWideChar(CP_ACP, MB_ERR_INVALID_CHARS | MB_PRECOMPOSED,
+                        str, -1, wstr, wstr_size);
+# else
```

Comment:
Everything points toward that this #else should rather be #elif defined(__linux). As for #else presumably best one can do is to detect if locale is UTF-8 and display a warning. Question would be how. nl_langinfo(CODESET) is not part of C, so it would need to be detected separately. Alternatively one can check if mbstowcs("\xc3\xaf") yields single wchar_t or pair...

Reply:
> detect if locale is UTF-8 and display a warning

Naturally in case it's not UTF-8.

Reply:
> Alternatively one can check if mbstowcs("\xc3\xaf") yields single wchar_t or pair...

Or bails out with error...

Just in case "everything points towards" means that so far I haven't found another Unix that would be as understanding as Linux...